062-support-tiered-storage.md

```diff
@@ -0,0 +1,117 @@
+#  Support tiered storage natively in Strimzi
+
+## Current situation
+
+Kafka introduced tiered storage feature since 3.6. The feature is in early stage and it’s nice to support tiered storage configuration natively via Strimzi.
+
+## Motivation
+
+Supporting tiered storage via Strimzi allow Strimzi users to enable tiered storage natively with Strimzi config. By allowing to config custom configuration for `RemoteStorageManager`, users are able to adopt tiered storage feature on Strimzi easily.
+
+For simplicity, the proposal doesn't add support to customize `RemoteLogMetadataManager`. If tiered storage is enabled, the default implementation class `TopicBasedRemoteLogMetadataManager` in Kafka will be used.
+
+This proposal focuses on tiered storage support with a custom plugin brought in by the user. Other proposals in the future might add some built-in support for selected storage implementations such as S3. It should also be noted that, it’s on the end-user to ensure that their tiered storage setup works with Strimzi; the operator itself is strictly responsible for pushing down the necessary config to the broker pods.
+
+## Proposal
+
+Add a `spec.kafka.tieredStorage` field in the `Kafka` CR, which would support configuration for all tiered storage related configuration.
+
+The `tieredStorage` config contains configuration for some selected configuration of tiered storage feature. Most importantly, the configuration for `RemoteStorageManager` is needed.
+
+```
+kafka:
+  tieredStorage:
+    type: custom
+    rsm.className:
+    rsm.classPath:
+    rsm.impl.prefix: rsm.config.
+```
+
+### RemoteStorageManager (RSM)
+
+The RSM config only supports custom type. The config allows users to specify the RSM class information.
+
+Because the `RemoteStorageManager` dependency is provided by Kafka plugin, the implementation could vary case by case. It’s important to allow user to pass in additional configuration values within the supplied prefix range. The custom configuration can be supplied via  `spec.kafka.config` with the predefined prefix.
+
+### RemoteLogMetadataManager (RLMM)
+
+For simplicity, the proposal doesn't add customization for RemoteLogMetadataManager. If tiered storage is enabled, the default implementation using `TopicBasedRemoteLogMetadataManager` will be used, and corresponding configuration will be appended. 
+
+Since the default RLMM will create an internal Kafka client to talk to the server, there might be some additional configuration needed, like the SSL related setting. Those config can be put in `.spec.kafka.config` directly using the `rlmm.config.` prefix. Example can be found below.
+
+### Example configuration
+
+The below config define a sample configuration for tiered storage setup, using a `custom` RSM type.
+```
+kafka:
```

Comment:
Thinking a bit more about this API ... what if we used something like this:

Thinking 

```yaml
kafka:
  tieredStorage:
    type: customRemoteStorageManager
    className: com.example.kafka.tiered.storage.s3.S3RemoteStorageManager
    classPath:  /opt/kafka/plugins/tiered-storage-s3/*
    config: 
      # A map with String keys and String values. Keys will be automatically prefixed with `rcm.config.`.
      storage.bucket.name: my-bucket
  config:
    ...
    rlmm.config.remote.log.metadata.topic.replication.factor: 1
```

* The `type: customRemoteStorageManager` should make it more clear that this uses Kafka's `RemoteLogMetadataManager` and allows us to add some other custom types including a custom `RemoteLogMetadataManager` in the future.
* I do not think we need the `rsm.` prefix as the type makes it more clear that we configure the RemoteStorageManager. But if you think it should be there, we should avoid the dot-notation and call it just `rsmClassName` & `rsmClassPath`.
* I would add there a `config` section that will be automatically prefixed with `rsm.config.` and that can be used for any configuration specific to the RemoteStorageManager implementation.

The basic RLLM configuration will be done by Strimzi and only some finetuning will be done in the `.spec.kafka.config` section. The configuration above will result in a Kafka broker configuration containing this:

```properties
# RLMM configuration generated by Strimzi
remote.log.storage.system.enable=true
remote.log.metadata.manager.class.name=org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManager
remote.log.metadata.manager.impl.prefix=rlmm.config.
remote.log.metadata.manager.listener.name=REPLICATION-9091
rlmm.config.remote.log.metadata.common.client.security.protocol=SSL
rlmm.config.remote.log.metadata.common.client.ssl.keystore.location=/tmp/kafka/cluster.keystore.p12
rlmm.config.remote.log.metadata.common.client.ssl.keystore.password=${CERTS_STORE_PASSWORD}
rlmm.config.remote.log.metadata.common.client.ssl.keystore.type=PKCS12
rlmm.config.remote.log.metadata.common.client.ssl.truststore.location=/tmp/kafka/cluster.truststore.p12
rlmm.config.remote.log.metadata.common.client.ssl.keystore.password=${CERTS_STORE_PASSWORD}
rlmm.config.remote.log.metadata.common.client.ssl.truststore.type=PKCS12

# Custom RLMM configs
rlmm.config.remote.log.metadata.topic.replication.factor=1

#RSM configs set by the operator and by the user
remote.log.storage.manager.class.path=/opt/kafka/plugins/tiered-storage-s3/*
remote.log.storage.manager.class.name=com.example.kafka.tiered.storage.s3.S3RemoteStorageManager
remote.log.storage.manager.impl.prefix=rsm.config.
rsm.config.storage.bucket.name=my-bucket
```

That looks like something that matches the existing Strimzi APIs and allows to configure what is needed? WDYT?

Reply:
Thanks for the suggestion! Updated with API, have some minor update for the rlmm security config because those are actually fall under "custom RLMM config section".